cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableRegistry.java

```diff
@@ -110,17 +93,18 @@ public void initialize() throws IOException {
   public void registerSpecification(StructuredTableSpecification specification) throws TableAlreadyExistsException {
     LOG.debug("Registering table specification {}", specification);
     StructuredTableId tableId = specification.getTableId();
-    Transactionals.execute(
-      transactional,
-      context -> {
-        Table table = context.getDataset(ENTITY_REGISTRY, Collections.emptyMap());
-        byte[] serialized = table.get(getRowKeyBytes(tableId), SCHEMA_COL_BYTES);
-        if (serialized != null) {
-          throw new TableAlreadyExistsException(tableId);
-        }
-        serialized = Bytes.toBytes(GSON.toJson(specification));
-        table.put(getRowKeyBytes(tableId), SCHEMA_COL_BYTES, serialized);
-      }, TableAlreadyExistsException.class);
+    MetricsTable table = getRegistryTable();
+    try {
+      byte[] rowKeyBytes = getRowKeyBytes(tableId);
+      byte[] serialized = table.get(rowKeyBytes, SCHEMA_COL_BYTES);
+      if (serialized != null) {
+        throw new TableAlreadyExistsException(tableId);
+      }
+      serialized = Bytes.toBytes(GSON.toJson(specification));
```

Comment:
better use swap() to make sure there is no conflict

Reply:
+1